From 1e095a95e0b78b3627d287125b659e82084f599a Mon Sep 17 00:00:00 2001
From: George Zacharia <george.zcharia@gmail.com>
Date: Sun, 12 Apr 2020 00:25:09 +0530
Subject: [PATCH] Weather tile: Uncouple the text from image

The temperature string was overlayed into the weather condition image, I have not researched fully why it was done so.

Let's just uncouple the text and the image and set the text to a separate Textview
Change-Id: I7d712e92d8285c889efe8fc44d609feb3700716c
Change-Id: I7144cdaf06cbe7298cb69b496028ef68bed7cc0c
---
 .../res/layout/detailed_weather_view.xml      | 54 ++++++++++++++
 .../systemui/omni/DetailedWeatherView.java    | 74 ++++++++++---------
 2 files changed, 92 insertions(+), 36 deletions(-)

diff --git a/packages/SystemUI/res/layout/detailed_weather_view.xml b/packages/SystemUI/res/layout/detailed_weather_view.xml
index 3e46e4eaaa2..87b0f0ef0a6 100644
--- a/packages/SystemUI/res/layout/detailed_weather_view.xml
+++ b/packages/SystemUI/res/layout/detailed_weather_view.xml
@@ -77,8 +77,17 @@
                     android:layout_width="wrap_content"
                     android:layout_height="wrap_content"
                     android:layout_gravity="center_horizontal"
+                    android:paddingBottom="8dp"
                     android:scaleType="centerInside"/>
 
+                <TextView
+                    android:id="@+id/current_temp"
+                    android:layout_width="wrap_content"
+                    android:layout_height="wrap_content"
+                    android:layout_gravity="center_horizontal"
+                    android:textAppearance="@style/TextAppearance.StatusBar.Expanded.Weather.Detailed"
+                    android:singleLine="true" />
+
                 <TextView
                     android:id="@+id/current_text"
                     android:layout_width="wrap_content"
@@ -100,8 +109,17 @@
                     android:layout_width="wrap_content"
                     android:layout_height="wrap_content"
                     android:layout_gravity="center_horizontal"
+                    android:paddingBottom="8dp"
                     android:scaleType="centerInside"/>
 
+                <TextView
+                    android:id="@+id/forecast_temp_0"
+                    android:layout_width="wrap_content"
+                    android:layout_height="wrap_content"
+                    android:layout_gravity="center_horizontal"
+                    android:textAppearance="@style/TextAppearance.StatusBar.Expanded.Weather.Detailed"
+                    android:singleLine="true" />
+
                 <TextView
                     android:id="@+id/forecast_text_0"
                     android:layout_width="wrap_content"
@@ -123,8 +141,17 @@
                     android:layout_width="wrap_content"
                     android:layout_height="wrap_content"
                     android:layout_gravity="center_horizontal"
+                    android:paddingBottom="8dp"
                     android:scaleType="centerInside"/>
 
+                <TextView
+                    android:id="@+id/forecast_temp_1"
+                    android:layout_width="wrap_content"
+                    android:layout_height="wrap_content"
+                    android:layout_gravity="center_horizontal"
+                    android:textAppearance="@style/TextAppearance.StatusBar.Expanded.Weather.Detailed"
+                    android:singleLine="true" />
+
                 <TextView
                     android:id="@+id/forecast_text_1"
                     android:layout_width="wrap_content"
@@ -146,8 +173,17 @@
                     android:layout_width="wrap_content"
                     android:layout_height="wrap_content"
                     android:layout_gravity="center_horizontal"
+                    android:paddingBottom="8dp"
                     android:scaleType="centerInside"/>
 
+                <TextView
+                    android:id="@+id/forecast_temp_2"
+                    android:layout_width="wrap_content"
+                    android:layout_height="wrap_content"
+                    android:layout_gravity="center_horizontal"
+                    android:textAppearance="@style/TextAppearance.StatusBar.Expanded.Weather.Detailed"
+                    android:singleLine="true" />
+
                 <TextView
                     android:id="@+id/forecast_text_2"
                     android:layout_width="wrap_content"
@@ -169,8 +205,17 @@
                     android:layout_width="wrap_content"
                     android:layout_height="wrap_content"
                     android:layout_gravity="center_horizontal"
+                    android:paddingBottom="8dp"
                     android:scaleType="centerInside"/>
 
+               <TextView
+                    android:id="@+id/forecast_temp_3"
+                    android:layout_width="wrap_content"
+                    android:layout_height="wrap_content"
+                    android:layout_gravity="center_horizontal"
+                    android:textAppearance="@style/TextAppearance.StatusBar.Expanded.Weather.Detailed"
+                    android:singleLine="true" />
+
                 <TextView
                     android:id="@+id/forecast_text_3"
                     android:layout_width="wrap_content"
@@ -192,8 +237,17 @@
                     android:layout_width="wrap_content"
                     android:layout_height="wrap_content"
                     android:layout_gravity="center_horizontal"
+                    android:paddingBottom="8dp"
                     android:scaleType="centerInside"/>
 
+                <TextView
+                    android:id="@+id/forecast_temp_4"
+                    android:layout_width="wrap_content"
+                    android:layout_height="wrap_content"
+                    android:layout_gravity="center_horizontal"
+                    android:textAppearance="@style/TextAppearance.StatusBar.Expanded.Weather.Detailed"
+                    android:singleLine="true" />
+
                 <TextView
                     android:id="@+id/forecast_text_4"
                     android:layout_width="wrap_content"
diff --git a/packages/SystemUI/src/com/android/systemui/omni/DetailedWeatherView.java b/packages/SystemUI/src/com/android/systemui/omni/DetailedWeatherView.java
index 27654677fca..ec7715d1d42 100644
--- a/packages/SystemUI/src/com/android/systemui/omni/DetailedWeatherView.java
+++ b/packages/SystemUI/src/com/android/systemui/omni/DetailedWeatherView.java
@@ -69,6 +69,11 @@ public class DetailedWeatherView extends FrameLayout {
     private ImageView mForecastImage2;
     private ImageView mForecastImage3;
     private ImageView mForecastImage4;
+    private TextView mForecastTemp0;
+    private TextView mForecastTemp1;
+    private TextView mForecastTemp2;
+    private TextView mForecastTemp3;
+    private TextView mForecastTemp4;
     private TextView mForecastText0;
     private TextView mForecastText1;
     private TextView mForecastText2;
@@ -79,6 +84,7 @@ public class DetailedWeatherView extends FrameLayout {
     private boolean mWithBackgroundColor;
     private boolean mShowCurrent = true;
     private View mCurrentView;
+    private TextView mCurrentTemp;
     private TextView mCurrentText;
     private View mProgressContainer;
     private TextView mStatusMsg;
@@ -125,6 +131,11 @@ public class DetailedWeatherView extends FrameLayout {
         mForecastImage2  = (ImageView) findViewById(R.id.forecast_image_2);
         mForecastImage3  = (ImageView) findViewById(R.id.forecast_image_3);
         mForecastImage4  = (ImageView) findViewById(R.id.forecast_image_4);
+        mForecastTemp0 = (TextView) findViewById(R.id.forecast_temp_0);
+        mForecastTemp1 = (TextView) findViewById(R.id.forecast_temp_1);
+        mForecastTemp2 = (TextView) findViewById(R.id.forecast_temp_2);
+        mForecastTemp3 = (TextView) findViewById(R.id.forecast_temp_3);
+        mForecastTemp4 = (TextView) findViewById(R.id.forecast_temp_4);
         mForecastText0 = (TextView) findViewById(R.id.forecast_text_0);
         mForecastText1 = (TextView) findViewById(R.id.forecast_text_1);
         mForecastText2 = (TextView) findViewById(R.id.forecast_text_2);
@@ -132,6 +143,7 @@ public class DetailedWeatherView extends FrameLayout {
         mForecastText4 = (TextView) findViewById(R.id.forecast_text_4);
         mCurrentView = findViewById(R.id.current);
         mCurrentImage  = (ImageView) findViewById(R.id.current_image);
+        mCurrentTemp = (TextView) findViewById(R.id.current_temp);
         mCurrentText = (TextView) findViewById(R.id.current_text);
         mStatusMsg = (TextView) findViewById(R.id.status_msg);
         mEmptyView = findViewById(android.R.id.empty);
@@ -204,51 +216,52 @@ public class DetailedWeatherView extends FrameLayout {
         String dayShort = sdf.format(new Date(cal.getTimeInMillis()));
 
         Drawable d = mWeatherClient.getWeatherConditionImage(weatherData.forecasts.get(0).conditionCode);
-        d = overlay(mContext.getResources(), d, weatherData.forecasts.get(0).low, weatherData.forecasts.get(0).high,
-                weatherData.tempUnits);
+        d = overlay(mContext.getResources(), d);
         mForecastImage0.setImageDrawable(d);
+        mForecastTemp0.setText(formatTempString(weatherData.forecasts.get(0).low, weatherData.forecasts.get(0).high, weatherData.tempUnits));
         mForecastText0.setText(dayShort);
 
         cal.add(Calendar.DATE, 1);
         dayShort = sdf.format(new Date(cal.getTimeInMillis()));
 
         d = mWeatherClient.getWeatherConditionImage(weatherData.forecasts.get(1).conditionCode);
-        d = overlay(mContext.getResources(), d, weatherData.forecasts.get(1).low, weatherData.forecasts.get(1).high,
-                weatherData.tempUnits);
+        d = overlay(mContext.getResources(), d);
         mForecastImage1.setImageDrawable(d);
+        mForecastTemp1.setText(formatTempString(weatherData.forecasts.get(1).low, weatherData.forecasts.get(1).high, weatherData.tempUnits));
         mForecastText1.setText(dayShort);
 
         cal.add(Calendar.DATE, 1);
         dayShort = sdf.format(new Date(cal.getTimeInMillis()));
 
         d = mWeatherClient.getWeatherConditionImage(weatherData.forecasts.get(2).conditionCode);
-        d = overlay(mContext.getResources(), d, weatherData.forecasts.get(2).low, weatherData.forecasts.get(2).high,
-                weatherData.tempUnits);
+        d = overlay(mContext.getResources(), d);
         mForecastImage2.setImageDrawable(d);
+        mForecastTemp2.setText(formatTempString(weatherData.forecasts.get(2).low, weatherData.forecasts.get(2).high, weatherData.tempUnits));
         mForecastText2.setText(dayShort);
 
         cal.add(Calendar.DATE, 1);
         dayShort = sdf.format(new Date(cal.getTimeInMillis()));
 
         d = mWeatherClient.getWeatherConditionImage(weatherData.forecasts.get(3).conditionCode);
-        d = overlay(mContext.getResources(), d, weatherData.forecasts.get(3).low, weatherData.forecasts.get(3).high,
-                weatherData.tempUnits);
+        d = overlay(mContext.getResources(), d);
         mForecastImage3.setImageDrawable(d);
+        mForecastTemp3.setText(formatTempString(weatherData.forecasts.get(3).low, weatherData.forecasts.get(3).high, weatherData.tempUnits));
         mForecastText3.setText(dayShort);
 
         cal.add(Calendar.DATE, 1);
         dayShort = sdf.format(new Date(cal.getTimeInMillis()));
 
         d = mWeatherClient.getWeatherConditionImage(weatherData.forecasts.get(4).conditionCode);
-        d = overlay(mContext.getResources(), d, weatherData.forecasts.get(4).low, weatherData.forecasts.get(4).high,
-                weatherData.tempUnits);
+        d = overlay(mContext.getResources(), d);
         mForecastImage4.setImageDrawable(d);
+        mForecastTemp4.setText(formatTempString(weatherData.forecasts.get(4).low, weatherData.forecasts.get(4).high, weatherData.tempUnits));
         mForecastText4.setText(dayShort);
 
         if (mShowCurrent) {
             d = mWeatherClient.getWeatherConditionImage(weatherData.conditionCode);
-            d = overlay(mContext.getResources(), d, weatherData.temp, null, weatherData.tempUnits);
+            d = overlay(mContext.getResources(), d);
             mCurrentImage.setImageDrawable(d);
+            mCurrentTemp.setText(formatTempString(weatherData.temp, null, weatherData.tempUnits));
             mCurrentText.setText(mContext.getResources().getText(R.string.omnijaws_current_text));
         }
 
@@ -257,42 +270,21 @@ public class DetailedWeatherView extends FrameLayout {
         }
     }
 
-    private Drawable overlay(Resources resources, Drawable image, String min, String max, String tempUnits) {
+    private Drawable overlay(Resources resources, Drawable image) {
         if (image instanceof VectorDrawable) {
             image = applyTint(image);
         }
         final Canvas canvas = new Canvas();
         canvas.setDrawFilter(new PaintFlagsDrawFilter(Paint.ANTI_ALIAS_FLAG,
                 Paint.FILTER_BITMAP_FLAG));
-        final float density = resources.getDisplayMetrics().density;
-        final int footerHeight = Math.round(18 * density);
-        final int imageWidth = image.getIntrinsicWidth();
-        final int imageHeight = image.getIntrinsicHeight();
-        final TextPaint textPaint = new TextPaint(Paint.ANTI_ALIAS_FLAG);
-        Typeface font = Typeface.create("@*android:string/config_bodyFontFamily", Typeface.NORMAL);
-        textPaint.setTypeface(font);
-        textPaint.setColor(getTintColor());
-        textPaint.setTextAlign(Paint.Align.LEFT);
-        final int textSize= (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_SP, 14f, resources.getDisplayMetrics());
-        textPaint.setTextSize(textSize);
-        final int height = imageHeight + footerHeight;
-        final int width = imageWidth;
+        final int height = image.getIntrinsicHeight();
+        final int width = image.getIntrinsicWidth();
 
         final Bitmap bmp = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888);
         canvas.setBitmap(bmp);
-        image.setBounds(0, 0, imageWidth, imageHeight);
+        image.setBounds(0, 0, width, height);
         image.draw(canvas);
 
-        String str = null;
-        if (max != null) {
-            str = min +"/"+max + tempUnits;
-        } else {
-            str = min + tempUnits;
-        }
-        Rect bounds = new Rect();
-        textPaint.getTextBounds(str, 0, str.length(), bounds);
-        canvas.drawText(str, width / 2 - bounds.width() / 2, height - textSize / 2, textPaint);
-
         return new BitmapDrawable(resources, bmp);
     }
 
@@ -346,4 +338,14 @@ public class DetailedWeatherView extends FrameLayout {
     public void stopProgress() {
         mProgressContainer.setVisibility(View.GONE);
     }
+
+    private static String formatTempString(String min, String max, String tempUnits) {
+        String str = null;
+        if (max != null) {
+            str = min + "/" + max + tempUnits;
+        } else {
+            str = min + tempUnits;
+        }
+        return str;
+    }
 }
